src/services/blocks/trace/types.ts

```diff
@@ -0,0 +1,206 @@
+import { AccountInfo, Address } from '@polkadot/types/interfaces';
+import BN from 'bn.js';
+
+export interface StringValues {
+	key?: string;
+	method?: 'Put' | 'Get';
+	result?: string;
+	[i: string]: unknown;
+}
+
+export interface Data {
+	stringValues: StringValues;
+}
+
+export interface TraceEvent {
+	name: string;
+	parentId: number;
+	target: string;
+	data: Data;
+}
+
+/**
+ * Parsed event.
+ */
+export interface PEvent extends TraceEvent {
+	storagePath: KeyInfo;
+	parentSpanId?: ParentSpanId;
+	eventIndex: number;
+}
+
+/**
+ * Parsed event with info about the action group it belongs to.
+ */
+export interface PActionEvent extends PEvent {
+	phase: PhaseId;
+	primarySpanId: ParentSpanId;
+}
+
+/**
+ * Parsed event of a `Get` from the `AccountInfo` storage item. Has info about
+ * the action group it belongs to and the decoded `AccountInfo` storage item.
+ */
+export interface PAccountEvent extends PActionEvent {
+	accountInfo: AccountInfo;
+	address: Address;
+}
+
+export interface TraceSpan {
+	id: number;
+	name: string;
+	parentId: number;
+	target: string;
+	data: Data;
+}
+
+/**
+ * Span with direct descendants
+ */
+export interface SpanWithChildren extends TraceSpan {
+	/**
+	 * Ids of the child spans.
+	 */
+	children: number[];
+}
+
+export interface BlockTrace {
+	blockHash: string;
+	storageKeys: string;
+	tracingTargets: string;
+	events: TraceEvent[];
+	spans: TraceSpan[];
+}
+
+export function isBlockTrace(
+	thing: unknown
+): thing is { blockTrace: BlockTrace } {
+	return (
+		typeof (thing as { blockTrace: BlockTrace })?.blockTrace?.blockHash ===
+			'string' &&
+		typeof (thing as { blockTrace: BlockTrace })?.blockTrace?.tracingTargets ===
+			'string' &&
+		typeof (thing as { blockTrace: BlockTrace })?.blockTrace?.storageKeys ===
+			'string' &&
+		Array.isArray((thing as { blockTrace: BlockTrace })?.blockTrace?.events) &&
+		Array.isArray((thing as { blockTrace: BlockTrace })?.blockTrace?.spans)
```

Comment:
Repeating `(thing as { blockTrace: BlockTrace })?.blockTrace?` many times seems inefficient. Can't you

```
let bt_thing = (thing as { blockTrace: BlockTrace })?.blockTrace?;
return ['blockHash', 'tracingTargets', 'storageKeys'].all( member => typeof bt_thing[member] === 'string') && Array.isArray(… … …
```
?
(rewritten in actual js ofc)